### DISCUSSION TITLE
What is the logic of discovering cwe?

### DISCUSSION THREAD

**Discussion by Shmily8421** [fkie-cad/cwe_checker#296] [General]
As stated in the title

**Comment by Enkelmann** [comment]
Depends on the CWE and its corresponding CWE check. ;-) We employ different analysis techniques for different CWEs, ranging from simple heuristics to more complex data flow analyses. And since the cwe_checker is a research project, we are still improving the different checks and analysis techniques. If you are interested in a specific CWE check I can talk about the internals of that check.

**Reply by Shmily8421** [reply]
Thank you for your reply. But from the point of view of the whole project, its logic is to take the data generated by ghidra as IR, then use the `analysis` module to construct a unified graph structure, and use the `checker` module to check for differentiation. Do I understand it correctly?

**Reply by Enkelmann** [reply]
Yes. In addition to the creation of the control flow graph the analysis module also contains the `pointer_inference` module, which is a combined points-to and value set analysis. The CWE-specific checks are all contained in the `checker` module. With the notable exception of the memory check, which is currently directly integrated into the pointer inference module. But I am currently working on refactoring this part of the code so that the memory checks will also get their own modules inside the `checker` module.

**Reply by Shmily8421** [reply]
Mm-hmm. In fact, the most important thing in this project is the `analysis` module, because its degree determines the degree to which it can be analyzed later, and this module is very complex.

**Reply by Enkelmann** [reply]
Yes.

**Reply by Shmily8421** [reply]
Okay, thank u so mush!

**Reply by Shmily8421** [reply]
If you have time, can you explain the logic of the build graph? I really don't understand how this graph is built.

**Reply by Enkelmann** [reply]
- For each basic block contained in the program we add two nodes (a block start node and a block end node) and an edge between them to the graph. Actually, every basic block node is labeled by the block ID and the sub ID in which it is contained. But this is an artifact from previous times where basic blocks could happen to be contained in more than one function, which is not longer the case.
- For direct jumps and indirect jumps where Ghidra gives us a list of possible jump targets we add an edge from the block end node of the source of the jump to the block start node of the target of the jump.
- For calls to external functions we add an edge from the block end node of the block containing the call to the block start node of the basic block that the call returns to. And the edge is labeled as an external call.
- For call to other functions inside the analyzed binary we add two nodes and a bunch of edges. This is actually described in the documentation comment of the `graph` module.
- Currently, we treat indirect calls as if they were calls to unknown external functions and add extern call edges for them. This is because we have not yet implemented an analysis identifying the targets of indirect calls.

Especially the addition of the call-specific nodes and edges can be a bit confusing to read, since we first have to collect all the call sites and return sites before we can add them. And one function may contain more than one return instruction, which further complicates things.
